simple hack to allow searching in theindex.html by matching word boundaries

timotheecour opened the pull request:
CI failure caused by https://github.com/timotheecour/Nim/issues/167

timotheecour committed:
simple hack to allow searching in theindex.html by matching word boundaries

Araq commented on lib/packages/docutils/rstgen.nim:
```diff
@@ -457,8 +457,9 @@ proc generateSymbolIndex(symbols: seq[IndexEntry]): string =
-                [keyword, cleanedKeyword])
+    let prefix = "-" # allows restricting search to match beginning of word, eg: `-split`
+    result.addf("<dt><a name=\"$3\" href=\"#$3\"><span>$1$2:</span></a></dt><dd><ul class=\"simple\">\n",
+                [prefix,keyword, cleanedKeyword])
```
1. The diff is ``<span>-$1</span>`` but you managed to make this much harder to see. ;-)
2. Is this an observable change?
3. Can't we do this in a better way? 
4. What problem does this solve?

timotheecour replied:
1. done, changed to small diff
2. yes, now top-level entries start with `-`, try this PR with `nim doc -r --project compiler/index` and open index, then enjoy the fact that searching for `-render` brings up only the symbols starting with `render` (as opposed to other symbols with render in the middle, or to random other text (in proc signatures etc) that has `render` in it) => 15 consecutive choices to choose from in your search instead of 79 non-consecutive choices
the `-` blends in well and doesn't look ugly (IANAD(esigner))
3. force users to use a browser plugin that supports regex search to search for `\brender` instead of `-render` or `ode:` instead of `ode\b`
=> that's not as effective because it would still not allow a way that only matches against symbols and not distractors; another way is to enhance dochack for this use case (can be done in future work, with other use cases like searching by signature, but that's much more involved; also dochack isn't enabled inside theindex right now)
4. allow searching for "just symbols" like the `renderModule` entry matching either at beginning (-render) or at end (odule:), without distractors found inside the text (`renderer: renderModule(n: PNode; infile, outfile: string; renderFlags: TRenderFlags = {}; fid = FileIndex(-1); conf: ConfigRef = nil)` :

```
-renderModule: <= the symbol i want to match 
renderer: renderModule(n: PNode; infile, outfile: string; renderFlags: TRenderFlags = {}; fid = FileIndex(-1); conf: ConfigRef = nil)`  <= the distractor
```

timotheecour replied:
@araq ping on this

timotheecour committed:
address comment

timotheecour committed:
fixup after rebase; fix {.base.} method warning

timotheecour commented:
ping @araq this feature is useful and saves time when searching
(just rebased to fix conflicts)

Araq commented:
But I don't like to **see** '-' prefixes in the produced documentation just for your searching conveniences. There should be a better way.

timotheecour committed:
make prefix invisible (white)

timotheecour committed:
fix tests

timotheecour commented:
PTAL
> But I don't like to see '-' prefixes in the produced documentation just for your searching conveniences. There should be a better way.

fixed by using `<span style="color: white">-</span>"`
they're now invisible, yet you can search in theindex by `render` (matches every substring) or `-render` (matches by beginning of word).
dark mode isn't affected

Araq commented:
Ok, now who knows about this search feature? For now it's for you only, you need to make it visible to others somehow. I mean, we don't want undocumented features for the pleasure of single individuals, do we...

timotheecour committed:
API symbol word boundaries are now: `#foo:#`; add documentation hint for this feature

timotheecour committed:
fix tests

timotheecour commented:
PTAL
* added a small hint in top to explain succintly
* changed to `#` since `-` had quite a few false positives for stdlib, whereas `#` only appears once
* also matched `#` at end (`:` matching at end has many false positives, this makes it more robust for suffix searches)
* not visually disruptive since these are hidden

## example
you can play with this if you want to see how it works before merging by (after re-compiling) running `nim doc -r --project lib/pure/os.nim`, opening theindex, and start typing for `write` (~136 hits) vs `#writ` (~12 hits, all of which are the API symbols starting with `writ`)
